crates/walrus-service/src/event/event_blob_downloader.rs

```diff
@@ -259,4 +220,112 @@ impl EventBlobDownloader {
 
         Ok(blobs)
     }
+
+    /// Reads the status of an event blob.
+    ///
+    /// When reading the first certified event blob, it may be the case that the event blob is
+    /// just certified, and the storage nodes may not know it yet. Therefore, upon encountering
+    /// a nonexistent blob, we will retry and wait for the blob to be certified.
+    /// For all the earlier blobs, they must have been seen by the storage nodes, since
+    /// otherwise, no new certified event blobs would be created.
+    async fn read_event_blob_status(
+        &self,
+        event_blob_id: BlobId,
+        reading_first_event_blob: bool,
+        start_time: Instant,
+    ) -> Result<BlobStatus> {
+        let blob_status = loop {
+            let blob_status = match self
+                .walrus_client
+                .get_blob_status_with_retries(&event_blob_id, &self.sui_read_client)
+                .await
+            {
+                Ok(blob_status) => blob_status,
+                Err(err) if matches!(err.kind(), ClientErrorKind::BlobIdDoesNotExist) => {
+                    BlobStatus::Nonexistent
+                }
+                Err(err) => return Err(err.into()),
+            };
+
+            if blob_status == BlobStatus::Nonexistent && reading_first_event_blob {
+                tracing::debug!(
+                    "reading first certified event blob {} encountered a nonexistent \
+                blob, waiting for it to be certified",
+                    event_blob_id
+                );
+
+                if start_time.elapsed() > Duration::from_secs(30) {
+                    tracing::warn!(
```

Comment:
It'd be nice to add a const for some of the magic numbers here: `FIRST_BLOB_CERTIFICATION_TIMEOUT`, etc